### ISSUE TITLE
Generating SQL from dplyr with latest package versions orders of magnitude slower compared to older versions.

### ISSUE THREAD

**Issue by jozefhajnala** [sparklyr/sparklyr#3009]
It seems that the upgrade to the current package versions (`sparklyr`, `dplyr`, `dbplyr`) significantly (up to about 1000-fold) slows down the conversion of dplyr code to SQL that is sent to Spark for execution. This might seem immaterial in small interactive use, but with larger scale applications the overhead can become quite significant, where minutes are spent just translating dplyr code to SQL (where it would take milliseconds with the older versions).

# Reproducible example

Because of the complexity of the setup, here is a Docker-based script (tested on ubuntu) that can help reproduce the results. It is mostly a few very simple mutate chains. Unfortunately my knowledge of the underlying dispatch is not deep enough to pinpoint the culprit, but hopefully this will help investigating:

```bash
# First, make the benchmarking R script
echo "library(dplyr)
library(dbplyr)
library(sparklyr)
library(microbenchmark)
sc <- spark_connect('local')
cars <- copy_to(sc, df = datasets::mtcars, name = 'cars')
mb <- microbenchmark(
 unit = 'ms',
 times = 20,
 one = {
   res <- cars %>%
     mutate(a = 1)
 },
 three = {
   res <- cars %>%
     mutate(a = 1) %>%
     mutate(b = 2) %>%
     mutate(c = 3)
 },
 seven = {
   res <- cars %>%
     mutate(a = 1) %>%
     mutate(b = 2) %>%
     mutate(c = 3) %>%
     mutate(d = 4) %>%
     mutate(e = 5) %>%
     mutate(f = 6) %>%
     mutate(g = 7)
 },
 join = {
   res <- cars %>%
     select(-gear) %>%
     left_join(cars, by = 'cyl') %>%
     left_join(cars, by = 'gear')
 }
)
sessionInfo()
print(mb)
" > ./testscript.R


# Make an R script that upgrades relevant packages with latest binaries
echo "
options(repos = 'https://packagemanager.rstudio.com/all/__linux__/focal/latest')
install.packages(c('dplyr', 'dbplyr', 'tidyr', 'sparklyr', 'DBI', 'microbenchmark'))
" > ./installnewversions.R


# Run the benchmark with old package versions
docker run --rm -it \
  -v $(pwd)/testscript.R:/testscript.R \
  jozefhajnala/jozefio \
  /bin/bash -c "Rscript /testscript.R"

# Run the benchmark with latest package versions
docker run --rm -it \
  -v $(pwd)/testscript.R:/testscript.R \
  -v $(pwd)/installnewversions.R:/installnewversions.R \
  jozefhajnala/jozefio \
  /bin/bash -c "Rscript /installnewversions.R; Rscript /testscript.R"

# Cleanup
rm ./testscript.R
rm ./installnewversions.R
```

# Comparison results: 

## sparklyr_1.6.2,      dbplyr_2.1.0,        dplyr_1.0.5

```r
Unit: milliseconds
  expr       min        lq      mean    median        uq       max neval
   one  123.2602  154.3882  166.1230  157.3692  180.3262  213.9203    20
 three  664.5544  752.9070  772.7108  782.5937  798.2484  826.2670    20
 seven 2713.4621 2869.5349 2909.6972 2907.5903 2967.2505 3141.2271    20
  join  476.0601  494.1815  535.9080  508.7171  522.1112 1013.5114    20
````

## sparklyr_1.0.4,       dbplyr_1.4.2,        dplyr_0.8.3

```r
Unit: milliseconds
  expr      min       lq      mean   median        uq       max neval
   one 0.247967 0.268820 0.3057914 0.282941 0.3243305  0.521591    20
 three 1.693421 1.798379 2.7930008 1.825486 2.0409255 17.755982    20
 seven 6.350737 6.640808 7.0498743 6.824367 6.9105510  8.551317    20
  join 1.672493 1.818017 2.7935824 1.892949 2.0847110 19.483860    20
````

## TL;DR:

It can take up to 3 seconds to create the SQL for a simple `mutate` , whereas with the previous versions this took < 10 ms.

Is this performance penalty expected? Is there any way to circumvent this?

**Comment by yitao-li** [comment]
@jozefhajnala The only change in sparklyr 1.6.x is switching from dbplyr backend API edition 1 to edition 2 (see https://dbplyr.tidyverse.org/articles/backend-2.html for technical detail about what this transition entails).

This change should not have much effect on the resulting Spark SQL queries, but it does cause a different set of S3 methods to be called while generating the queries. If you believe time taken to generate such queries is a real perf bottleneck, you can set `options(sparklyr.dbplyr.edition = 1L)` to use the old dbplyr API with sparklyr 1.6.x

**Comment by nathaneastwood** [comment]
@yitao-li could you possibly clear something up for me please?

> * dbplyr 2.2.0 (to be released >= 12 months after dplyr 2.0.0) removes the old interface so user must upgrade backends.

Does this mean that when version 2.2.0 is released we will no longer be able to use "edition 1" if we install this version of dbplyr?

**Comment by yitao-li** [comment]
@nathaneastwood That is my interpretation as well. Sparklyr is good for now in relation to this dbplyr API change (in fact it is compatible with both editions of dbplyr API at the moment), but if you have other dbplyr-dependent R packages, then the authors of those R packages should ideally make changes outlined in https://dbplyr.tidyverse.org/articles/backend-2.html for their packages to be compatible with future versions of dbplyr.

**Comment by jozefhajnala** [comment]
Hi @yitao-li, thanks for the suggestion. I tried with the option you suggested:
```
options(sparklyr.dbplyr.edition = 1L)
````

Also with setting the env. var: 

```
export DBPLYR_API_EDITION=1;
```

And also with passing directly to the config

```
config <- spark_config()
config[['sparklyr.dbplyr.edition']] = 1L
sc <- spark_connect('local', config = config)
```

But no luck with either option, the performance is still heavily impacted.

I was able to trace is down to the following: 

With the older version of the packages, all the translation happens within R, with no communication to the Spark instance. However, with the new packages, when using 

```
options(sparklyr.log.invoke = "cat")
``` 

We can see that there is a lot of invocations going on, which is what seems to cause that large performance overhead when looking at profiling. The origin of the invokes seems to be that `simulate_vars` which is called within `sparklyr:::mutate.tbl_spark` uses `sdf_schema`, which is the source of the invocations:

```
> res <- dplyr::mutate(cars, a = 1)
Invoking sql
Invoking schema
Invoking fields
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking dataType
Invoking toString
Invoking name
Invoking sql
Invoking columns
```

Is this the intended behavior?

**Comment by yitao-li** [comment]
@jozefhajnala Yes that was intended.

There is a method called sdf_schema() that translates the type of each Spark column into an equivalent R type. Currently the number of spark_invoke()s from this is same as number of columns of the Spark dataframe. I guess it can be optimized to process all columns with only 1 spark_invoke() instead.

**Comment by yitao-li** [comment]
Update: the change in https://github.com/sparklyr/sparklyr/pull/3013 should help: it reduces the 3x `spark_invoke()`s per column down to only 1x `spark_invoke()` for the entire Spark dataframe.

**Comment by jozefhajnala** [comment]
Thanks, there is a significant improvement over the current CRAN versions (although still much slower compared to the "legacy" versions):
 
![image](https://user-images.githubusercontent.com/23148397/113770742-d13c3c00-9722-11eb-890d-c2134419141e.png)

In case you are interested I automated the benchmark via GitHub actions:
https://github.com/jozefhajnala/sparklyr_api_benchmarks/runs/2281392421?check_suite_focus=true

**Comment by yitao-li** [comment]
@jozefhajnala Thanks so much for writing the benchmark plus suggesting sparklyr.log.invoke = "cat" for debugging! I think there are still some more places where the extra spark_invoke()s can be optimized away BTW. I'll keep you posted in https://github.com/sparklyr/sparklyr/issues/3011
